### PR TITLE
[Bugfix][Perf] mimo_audio: fix per-request speech-code routing under continuous batching

### PR DESCRIPTION
**Pull request by JuanPZuluaga**
## Purpose

`MiMoAudioLLMForConditionalGeneration.forward()` performed ~7 device-to-host syncs per decode frame at batch size 1, growing by 6 syncs per additional request: each request read its CSR boundaries via `int(query_start_loc[i].item())` (twice, in two separate loops) and evaluated `bool(tensor)` empty-token checks individually, and the fallback `query_start_loc` tensor was built (H2D copy) even when unused. Each sync stalls the CPU behind the GPU stream om the per-frame decode loop.

This change collapses the per-frame sync count to a constant 3, independent of `num_reqs`:
- one `.tolist()` for the request boundaries,
- one `.cpu()` copy of the input empty-token mask,
- one `.cpu()` copy of the predicted-token empty mask.

## Changes

- `forward()`: materialize `query_start_loc` on the host once via a single `.tolist()`;   both downstream loops index the Python list instead of calling `.item()` per boundary   per request.
- `forward()`: build the fallback `query_start_loc` tensor only when `attn_metadata` is  absent, removing an unconditional per-step H2D copy.
- `_collect_merge_mm_embedding_info()`: take `query_start_loc_list: list[int]` and batch  the per-request `<|empty|>` check into one host copy of the mask instead of one `bool(tensor)` sync per request.
- Speech-token block: one host copy of the `next_ids == empty` mask replaces the `.any()` sync plus a per-request `bool()` sync; the `.any()` gate and all per-request checks read the CPU copy.

## Numbers

Correctness fix, with `vllm serve XiaomiMiMo/MiMo-Audio-7B-Instruct --omni` (async-chunk, 2 stages on one RTX 6000 Ada), N concurrent `/v1/chat/completions` speech requests:

| concurrency | main (ok/total) | with fix |
|---|---|---|
| 1 | 6/6 | 6/6 (e2e p50 2118 → 2120 ms, unchanged) |
| 4 | 9/24 | **24/24** |
| 8 | 11/48 | **48/48** |

main logs 104 AttributeError 500s + 209 ragged payload-extraction errors across the runs; with the fix, zero of both. Without the fix mimo_audio is effectively unusable at c>1.

also with nsys check: (eager, byte-identical output): at concurrency with more than 4,  with audio generation, −13.7% D2H copies / −8.8% stream syncs / −6.3% kernel launches per request; flat at bs=1.

## Test Plan

**vLLM Version:**

**vLLM-Omni Commit:**

## Test Result

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by linyueqian** [comment]
@qibaoyuan ptal

**Comment by qibaoyuan** [comment]
Could you add benchmark numbers to the Test Result section? Specifically:

- Decode step average latency at `bs=1, 4, 8` (before vs. after this change)
- End-to-end latency for one full audio generation

File: vllm_omni/model_executor/models/mimo_audio/mimo_audio_llm.py [hunk 1]
```diff
@@ -897,20 +897,24 @@ def _collect_merge_mm_embedding_info(
...
+        query_start_loc_list: list[int],
         kwargs: dict,
     ) -> tuple[dict[str, any], dict]:
         has_merge_mm_embedding = False
         merge_mm_embedding_info: dict[str, any] = {}
         seq_len = input_ids.shape[1] if input_ids.ndim == 2 else input_ids.shape[0]
 
+        # Copy the empty-token mask to host once; per-request checks below then
+        # avoid a GPU sync each.
+        input_ids_is_empty_cpu = (input_ids.reshape(-1) == self.empty_token_id).cpu()
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/mimo_audio/mimo_audio_llm.py]: I think this copies the whole flattened `input_ids` mask to host on every forward, but the loop below only reads it at request starts and only when `seq_len_by_req == 1`. On a big prefill that's a full-sequence D2H the old code never paid (the empty check only fired for single-token reqs). Could we gather just the start positions, or gate this behind a decode/`num_reqs` check, so we don't move the entire sequence each step? 🤔

File: vllm_omni/model_executor/models/mimo_audio/mimo_audio_llm.py [hunk 2]
```diff
@@ -897,20 +897,24 @@ def _collect_merge_mm_embedding_info(
...
         for req_idx, req_id in enumerate(request_ids):
-            query_start_loc_by_req = int(query_start_loc[req_idx].item())
-            query_end_loc_by_req = int(query_start_loc[req_idx + 1].item())
+            query_start_loc_by_req = query_start_loc_list[req_idx]
+            query_end_loc_by_req = query_start_loc_list[req_idx + 1]
             input_ids_by_req = input_ids[query_start_loc_by_req:query_end_loc_by_req]
             seq_len_by_req = input_ids_by_req.shape[1] if input_ids_by_req.ndim == 2 else input_ids_by_req.shape[0]
 
-            if seq_len_by_req == 1 and bool(input_ids_by_req == self.empty_token_id):
+            if seq_len_by_req == 1 and bool(input_ids_is_empty_cpu[query_start_loc_by_req]):
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/mimo_audio/mimo_audio_llm.py]: I want to double-check the axis here: `input_ids_is_empty_cpu` comes from `input_ids.reshape(-1)`, but two lines up `input_ids[query_start_loc_by_req:query_end_loc_by_req]` slices dim 0. When `input_ids` is 2D (the `ndim == 2` guards imply it can be), do the flat token offsets still line up with `query_start_loc_list`? If it's always 1D on this path we're fine, just want to be sure we're indexing the same axis. :)

File: vllm_omni/model_executor/models/mimo_audio/mimo_audio_llm.py [hunk 3]
```diff
@@ -1036,12 +1040,14 @@ def forward(
...
         _forward_context = get_forward_context()
-        _default_query_start_loc = torch.tensor([0, input_ids.shape[-1]], device=input_ids.device)
-        query_start_loc = (
-            next(iter(_forward_context.attn_metadata.values())).query_start_loc
-            if _forward_context.attn_metadata is not None
-            else _default_query_start_loc
-        )
+        if _forward_context.attn_metadata is not None:
+            query_start_loc = next(iter(_forward_context.attn_metadata.values())).query_start_loc
+        else:
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/mimo_audio/mimo_audio_llm.py:1052]: Since we immediately `.tolist()` right below, this fallback allocates a device tensor just to sync it straight back to host — a little against the grain of the PR. I'd build the host list directly in this branch, e.g. `query_start_loc_list = [0, input_ids.shape[-1]]`, and only `.tolist()` in the attn-metadata branch.

File: vllm_omni/model_executor/models/mimo_audio/mimo_audio_llm.py (hunk 1 above)
Comment (JuanPZuluaga) [reply vllm_omni/model_executor/models/mimo_audio/mimo_audio_llm.py]: Fixed now here, gathers only the per-request start positions instead of copying the whole flattened `input_ids` mask each step.

File: vllm_omni/model_executor/models/mimo_audio/mimo_audio_llm.py (hunk 2 above)
Comment (JuanPZuluaga) [reply vllm_omni/model_executor/models/mimo_audio/mimo_audio_llm.py]: correct,`input_ids` is 1-D on this path, so the flat offsets line up with `query_start_loc`.

**Comment by JuanPZuluaga** [comment]
@qibaoyuan, i did a small latency test, the decode per step latency is flat at bs=1 (~35 ms, within noise) so not much gains here.

The change is a per-step sync-count reduction (−13.7% D2H at bs≥4) that doesn't move wall-clock on Ada (my current machine).

**Comment by qibaoyuan** [comment]
LGTM!

**Review by linyueqian — APPROVED** [review]
